### PR TITLE
Add a separate service for ws traffic

### PR DESCRIPTION
**Pull request by mwiencek**
Luks suggested this in #musicbrainz-devel and it seems like a very sound idea - I split the current worker procs 10:40 with 10 going to / and 40 going to /ws/. Might need tweaking.

This is probably something that'd happen eventually anyway if we split out musicbrainz-ws from musicbrainz-server.

### CODE REVIEW

**Comment by mwiencek** [comment]
Open question...are there any problems that'd result from this with the in-memory caches not being shared? (If so, maybe time to remove it? :P)

**Comment by ianmcorvidae** [comment]
It'll mean we don't get as much benefit from said caches, but I'm not sure we're getting much as it is. I'm not entirely sure they were per master process before either, they may have been per *worker*, which would mean this is no worse.

I commented on this a bit to you privately, but to repeat here: this invalidates my nginx tweaks, since it's sort of an incompatible method (that split/rebalanced by servers, where this does by processes on each server); this is probably more forward-thinking though. 40/10 is definitely weighting heavily toward the site, which in practice is probably more like 5% or less of our requests. Whether that's a problem I'm not sure, or how heavily the weighting might cause problems with increasing 502s in the WS.

I think my feeling is we may as well go with it, though, and see how things pan out with it. The ticket for this is probably comes down to MBH-407, also.

File: musicbrainz-server/templates/default/sv-musicbrainz-ws-run.erb [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+exec 2>&1
+set -e
+cd mb_server
+umask 022
+exec \
+    env MUSICBRAINZ_USE_PROXY=1 starman -Ilib \
+        --preload-app \
+        -E deployment \
+        --listen musicbrainz-ws.socket \
+        --workers <%= @variables[:nproc] || 5 %> \
```
Comment (ianmcorvidae) [inline musicbrainz-server/templates/default/sv-musicbrainz-ws-run.erb]: This should be nproc-ws, right?

Comment (ianmcorvidae) [inline musicbrainz-server/templates/default/sv-musicbrainz-ws-run.erb]: er, no, maybe notu. forgot about the rebinding thing that it does I guess

File: musicbrainz-server/recipes/production.rb [hunk 2]
```diff
@@ -15,7 +15,17 @@
...
+  variables :nproc => node['musicbrainz-server']['nproc-website']
+  action [:enable,:start]
+  subscribes :restart, "git[/home/musicbrainz/musicbrainz-server]"
+  subscribes :restart, "template[/home/musicbrainz/musicbrainz-server/lib/DBDefs.pm]"
+  log true
+end
+
+daemontools_service "musicbrainz-ws" do
+  directory "/home/musicbrainz/svc-musicbrainz-ws"
+  template "musicbrainz-server"
```
Comment (ianmcorvidae) [inline musicbrainz-server/recipes/production.rb]: I do think this needs changing, to 'musicbrainz-ws', so it finds the right templates. Unless, of course, you just want to have one set and change the socket to be a configurable variable, which I think would work too?

Comment (mwiencek) [inline musicbrainz-server/recipes/production.rb]: Hm, a separate one probably makes sense. I don't understand the mechanism by which it finds the `sv-*.erb` files in musicbrainz-server/templates/default/, would changing this to "musicbrainz-ws" do that automatically?

Comment (mwiencek) [inline musicbrainz-server/recipes/production.rb]: Ah, I didn't have the daemontools submodule checked out, so grep wasn't finding where it did that.

File: musicbrainz-server/templates/default/sv-musicbrainz-ws-run.erb (hunk 1 above)
Comment (mwiencek) [inline musicbrainz-server/templates/default/sv-musicbrainz-ws-run.erb]: Yeah, from what I understand the `variables :nproc => ` thing in the service description rebinds it here.

**Comment by mwiencek** [comment]
I wondered about the minimum number of worker procs the website would need (starman defaults to 10), and whether the frontends plus totoro would able to handle an increase in workers (e.g. leave 50 for the ws + an additional 10 for the site, since the latter would be low-traffic), or how we arrived at the current number. If that'd increase connections to totoro, it'd be good to deploy the pgbouncer stuff first. Or if it'd just cause more requests to get backed up, then wouldn't be a good idea at all. To me it's still unclear if the workers processes are being exhausted, or the socket backlog.

**Comment by ianmcorvidae** [comment]
It wouldn't increase the number of connections to postgres, it might to pgbouncer. That's the point of pgbouncer though, to allow there to be more connections to it than to postgres and pool them. So it's fine on that count.

I think the current number of worker processes was semi-experimentally-determined. I think making it 50/10 would be safer as a small step forward with this patch, and still fine -- the main thing that concerned me was reducing *anything* about the webservice's allocation, which would at present be something like 48 processes. Bringing lolo in is still the better thing to do, so hopefully they'll actually respond to that ticket sometime soon (why is it always the weekend that this stuff happens...)

**Comment by mwiencek** [comment]
Okay, updated the server-configs PR to to 50/10.

**Comment by mwiencek** [comment]
I realized /ws/js should be included with website traffic, so I changed the location directive.

**Comment by ianmcorvidae** [comment]
:shipit:

**Pull request merged**